Added support for VM delete

VojkoR opened the pull request:
Added support for VM delete

/cc @miha-plesko

VojkoR commented:
Use "Power -> Delete" (shown in image) to delete VM.

![how-to-delete-vm](https://user-images.githubusercontent.com/8434824/36377495-aaf7e9ac-1577-11e8-80e7-fc0cb510145c.png)

miha-plesko commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,15 @@
+  def raw_destroy
+    raise "VM has no #{ui_lookup(:table => "ext_management_systems")}, unable to destroy VM" unless ext_management_system
+    with_provider_object(&:undeploy)
+
```
Nitpicking, but could you please remove this empty line?

VojkoR replied:
Empty line removed

miha-plesko commented:
@miq-bot assign @agrare 
@miq-bot add_label enhancement,gaprindashvili/yes

agrare commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,14 @@
+  def raw_destroy
+    raise "VM has no #{ui_lookup(:table => "ext_management_systems")}, unable to destroy VM" unless ext_management_system
+    with_provider_object(&:undeploy)
+    ext_management_system.with_provider_connection do |service|
```
Since `with_provider_object` calls `with_provider_connection` can you do the `:undeploy` in this block so you only have to connect once?

gberginc replied:
@VojkoR I believe what @agrare is suggesting is not to use the `with_provider_object` to undeploy but rather use it like this

```
ext_management_system.with_provider_connection do |service|
  service.post_undeploy_vapp(ems_ref)
  service.delete_vapp(ems_ref)
end
```

https://github.com/xlab-si/fog-vcloud-director/blob/master/lib/fog/vcloud_director/requests/compute/post_undeploy_vapp.rb#L36

agrare commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,14 @@
+    ext_management_system.with_provider_connection do |service|
+      service.delete_vapp(ems_ref)
+    end
+    update_attributes!(:raw_power_state => "DELETED")
```
Is `DELETED` a valid power state?  How about "off" since it is a valid [vcloud_director state](https://github.com/xlab-si/fog-vcloud-director/blob/master/lib/fog/vcloud_director/parsers/compute/vm.rb#L70) and also recognized by the [vcloud provider](https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/cloud_manager/vm.rb#L9-L15)

VojkoR commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,23 @@
+    raise "VM has no #{ui_lookup(:table => "ext_management_systems")}, unable to destroy VM" unless ext_management_system
+    ext_management_system.with_provider_connection do |service|
+      begin
+        response = service.post_undeploy_vapp(ems_ref, :UndeployPowerAction => 'shutdown')
```
Before delete action is called we need to stop VM. In case VM is already stopped action will rise an error. To successfully delete VM we are catching this error and just log the message what happened.

VojkoR commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,23 @@
+      ensure
+        response = service.delete_vapp(ems_ref)
+        service.process_task(response.body)
+        update_attributes!(:raw_power_state => "terminated")
```
I think correct power state for deleted VM should be "terminated". Maybe I should set power state to  "unknown" instead of "terminated"?

VojkoR commented:
Hi @agrare I changed functionality as requested. I added some extra code, to correctly implement DELETE action. I wrote some comments for the code that was added and why it was added.

agrare commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,23 @@
+      begin
+        response = service.post_undeploy_vapp(ems_ref, :UndeployPowerAction => 'shutdown')
+        service.process_task(response.body)
+      rescue => err
```
Can you rescue the exact exception that is raised if the VM is already stopped?

agrare commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,23 @@
+      rescue => err
+        $vcloud_log.debug("vm=[#{name}], error: VM not running! Ignoring error: #{err}")
+      ensure
+        response = service.delete_vapp(ems_ref)
```
Don't like doing this in the ensure block assuming that you're rescuing a "vm already powered off" exception because it could be something completely different and we'll still try to do this.
I'd rather rescue the exact exception then do this outside the begin/rescue block like:

```
begin
  service.post_undeploy_vapp(ems_ref, :UndeployPowerAction => 'shutdown')
rescue Vcloud::VmAlreadyPoweredOff /* not sure what this exception actually is, just for demonstration */
  # this is expected if the vapp is already off
end

service.process_task(response.body)
update_attributes!(:raw_power_state => "terminated")
```

gberginc commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,20 @@
+
+  included do
+    supports :terminate do
+      unsupported_reason_add(:terminate, "The VM is powered on") unless current_state == "off"
```
@agrare we have modified the code to resemble the behaviour of VCD. Now, it is not possible to delete a VM that is running (powered on). To this end we had to redefine the `supports_terminate?` by looking at the `current_state` of the VM.

agrare replied:
Ah I like that, we actually have a helper method for this that you can use, https://github.com/ManageIQ/manageiq/blob/master/app/models/vm/operations/power.rb#L22-L24

gberginc replied:
Thanks @agrare. I’ll make another PR to use this.

gberginc commented on app/models/manageiq/providers/vmware/cloud_manager/vm/operations.rb:
```diff
@@ -1,3 +1,20 @@
+  def raw_destroy
+    raise "VM has no #{ui_lookup(:table => "ext_management_systems")}, unable to destroy VM" unless ext_management_system
+    ext_management_system.with_provider_connection do |service|
+      response = service.delete_vapp(ems_ref)
```
Instead of checking for exceptions we are simply passing to the underlying provider connection. Looking at OpenStack and Amazon, this is exactly how these two are handled. In case an exception occurs it will be propagated.

agrare replied:
:+1: perfect

agrare commented:
This looks good, can you send a followup to use that `vm_powered_on?` helper?  I don't want to hold this up for that though.

miha-plesko commented:
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1550841

@simaishi ^

miha-plesko commented:
@agrare apologies for delay, but here comes the followup PR: https://github.com/ManageIQ/manageiq-providers-vmware/pull/204

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit fa49cf6bbc6e66d5373f03ed62bc1e37f6127ec9
Author: Adam Grare <agrare@redhat.com>
Date:   Fri Feb 23 09:34:48 2018 -0500

    Merge pull request #184 from VojkoR/destroy_vm_for_vmware_provider
    
    Added support for VM delete
    (cherry picked from commit 1dfa303668712cc1ee028d640cfbbaed734ca5f3)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1552683
```
